ggml/src/ggml-cpu/ggml-cpu.c

```diff
@@ -7421,6 +7381,17 @@ static void ggml_compute_forward_mul_mat_one_chunk(
     }
 }
 
+// move to ggml-cpu-traits...
+static const struct ggml_cpu_tensor_traits* ggml_cpu_get_tensor_traits(
+        const struct ggml_tensor * src0)
+{
```

Comment:
Is it possible to have here a "src0->extra" that is not part of the CPU backend?
ie: can the ggml_compute_zzzz be call with weight that is part of an other backend/device/backend?

Reply:
I don't see the point. As I already told, the way this is intended to be handled is with `ggml_backend_sched`.

Reply:
I think my question was not clear.
In the ggml_compute_forward_mul_mat we test that the buffer is a "aarch64" before use extra.
Can I remove this test an be sure that if extra exist at this point it have be set by the CPU backend?

Is this test there to differentiate between the (future) different cpu extra buffer?

If so if we can have the same struct (or base class...) for all cpu buffer type then we can remove this test.

My question was not to make it possible, but to see if we can simplify this function.

Reply:
There are some reasons reasons why we should not do this:
- The RPC backend does not support backends that use extras. This is not completely unavoidable, but currently this is done because the cost of calling `init_tensor` on every tensor is too high since it requires a round trip between the server and the client, so the RPC backend skips these calls.
- The CPU backend should be able to use buffers of iGPU backends that do not modify the weights to avoid extra copies. For example, the CPU backend can use Metal backend buffers since it uses host buffers. Making it require an extra would break that.
- And the same applies the other way around. The BLAS backend can use CPU buffers without copies because it can assume that the tensors stored in the default CPU buffer are in standard ggml layout. If we just use one buffer type and use the extras, this would no longer be possible.

So I think it is better to keep tensor conversions in different buffer types.

Reply:
> So I think it is better to keep tensor conversions in different buffer types.

Yes, I don't what to remove the fact that the tensor is in the "aarch64" backend, I just want to know if I can simplify this function like that:

```c++
static const struct ggml_cpu_tensor_traits* ggml_cpu_get_tensor_traits(const struct ggml_tensor * src0) {
    if (src0->extra != NULL) {
        return (struct ggml_cpu_tensor_traits*)src0->extra;
    }
    return NULL;
}
```

For what I see it is possible here, but I may have missed something.

Reply:
No, you should check the buffer type before using the extra because other backends that use host memory may want to set an extra. The CPU backend can use all buffer types that return true to `is_host`, and the only requirements for a buffer to be considered a "host buffer" is that `tensor->data` points to an address in system memory, and the tensor data is stored in standard ggml layout.

Reply:
> No, you should check the buffer type before using the extra because other backends that use host memory may want to set an extra. The CPU backend can use all buffer types that return true to `is_host`, and the only requirements for a buffer to be considered a "host buffer" is that `tensor->data` points to an address in system memory, and the tensor data is stored in standard ggml layout.
> 
> But you can rename the aarch64 buffer type to some generic name like "ggml_cpu_repack_buffer_type" and reuse it for the AMX or other repackings.

:+1:  OK I see. (did not read it before last reply) .